content/knowledge-base/guides/write-a-composition-function-in-python.md

```diff
@@ -0,0 +1,749 @@
+know about some other files in the template:
+
+* `function/main.py` runs the function. You don't need to edit `main.py`.
+* `Dockerfile` builds the function runtime. You don't need to edit `Dockerfile`.
+* The `package` directory contains metadata used to build the function package.
+
+{{<hint "tip">}}
+<!-- vale gitlab.FutureTense = NO -->
+<!--
+This tip talks about future plans for Crossplane.
+-->
+In v1.14 of the Crossplane CLI `crossplane beta xpkg init` just clones a
+template GitHub repository. In a future release the command will automate tasks
+like replacing the template name with the new function's name. See Crossplane
+issue [#4941](https://github.com/crossplane/crossplane/issues/4941) for details.
+<!-- vale gitlab.FutureTense = YES -->
+{{</hint>}}
+
+Edit `package/crossplane.yaml` to change the package's name before you start
+adding code. Name your package `function-xbuckets`.
+
+Some functions accept a configuration input. You configure the input when you
+write a Composition that uses the function. The `package/input` directory
+defines the OpenAPI schema for the a function's input. 
+
+The function in this guide doesn't accept an input. For this function you
+should delete the `package/input` directory. The
+[composition functions](https://docs.crossplane.io/latest/concepts/composition-functions)
+documentation explains more the input to a composition function.
+
+{{<hint "important">}}
+If you're writing a function that does use an input type, don't delete the
+`package/input` directory. Instead edit the file to be specific to your
+function.
+
+The kind `Input` is a placeholder value. The API group
+`template.fn.crossplane.io` is, too. Change the kind and API group to something
+meaningful to your function. Edit the `openAPIV3Schema` to represent your
+function's input schema.
+{{</hint>}}
+
+## Edit the template to add the function's logic
+
+You add your function's logic to the
+{{<hover label="hello-world" line="1">}}RunFunction{{</hover>}}
+method in `function/fn.py`. When you first open the file it contains a "hello
+world" function.
+
+```python {label="hello-world"}
+async def RunFunction(self, req: fnv1beta1.RunFunctionRequest, _: grpc.aio.ServicerContext) -> fnv1beta1.RunFunctionResponse:
+    log = self.log.bind(tag=req.meta.tag)
+    log.info("Running function")
+
+    rsp = response.to(req)
+
+    example = ""
+    if "example" in req.input:
+        example = req.input["example"]
+
+    # TODO: Add your function logic here!
+    response.normal(rsp, f"I was run with input {example}!")
+    log.info("I was run!", input=example)
+
+    return rsp
+```
+
+All Python composition functions have a `RunFunction` method. Crossplane passes
+everything the function needs to run in a
+{{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.
+The function tells Crossplane what resources it should compose by returning a
+{{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.
```

Comment:
```suggestion
All Python composition functions have a `RunFunction` method.  
Crossplane passes
everything the function needs to run in a
{{<hover label="hello-world" line="7">}}RunFunctionRequest{{</hover>}} object.  
The function tells Crossplane what resources to compose by returning a
{{<hover label="hello-world" line="22">}}RunFunctionResponse{{</hover>}} object.
```

Splitting across lines to improve readability.

I don't think your `{{<hover>}}` tags are matching what's intended.